tokio-trace/src/instrument.rs

```diff
@@ -53,18 +70,171 @@ impl<T: Sink> Sink for Instrumented<T> {
         &mut self,
         item: Self::SinkItem
     ) -> StartSend<Self::SinkItem, Self::SinkError> {
-        let span = &self.span;
+        let span = self.span.as_ref().cloned()
+            .expect("span never goes away for Sinks");
         let inner = &mut self.inner;
         span.enter(move || {
             inner.start_send(item)
         })
     }
 
     fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
-        let span = &self.span;
+        let span = self.span.as_ref().cloned()
+            .expect("span never goes away for Sinks");
         let inner = &mut self.inner;
         span.enter(move || {
             inner.poll_complete()
         })
     }
 }
+
+
+#[cfg(test)]
+mod tests {
+    use futures::{prelude::*, task, future, stream};
+    use ::{span, subscriber};
+    use super::*;
+
+    #[test]
+    fn future_enter_exit_is_reasonable() {
+        struct MyFuture {
+            polls: usize,
+        }
+
+        impl Future for MyFuture {
+            type Item = ();
+            type Error = ();
+            fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+                self.polls += 1;
+                if self.polls == 2 {
+                    Ok(Async::Ready(()))
+                } else {
+                    task::current().notify();
+                    Ok(Async::NotReady)
+                }
+            }
+        }
+        subscriber::mock()
+            .enter(
+                span::mock().named(Some("foo"))
```

Comment:
I really like this DSL!

Reply:
thanks! it's strongly inspired by the `h2` crate's mocking DSL: https://github.com/carllerche/h2/tree/master/tests/h2-support